Add flycheck support.

bsermons opened the pull request.

adamw commented:
Did you see ohai-emac's implementation? https://github.com/bodil/ohai-emacs/blob/master/modules/ohai-purescript.el#L67-L164

Not sure how it compares feature-wise but maybe these can be merged together :)

bsermons commented:
No, I was unaware of that implementation.  It does look more complete since this PR just gives the error messages at this time.  If @bodil wants to contribute it here, that would be preferred so not to duplicate effort.  If not, then I'll end up adding the suggestion feature since that was the initial plan anyway and I don't plan on using ohai since I already use spacemacs.

bodil commented:
Not sure whether it should be here or in a separate package? It's straightforward enough to add it here without making flycheck a dependency, but idk.

bsermons commented:
I don't see the problem with a flycheck dependency since this is basic IDE functionality and most people using this would have it already installed anyway I would imagine.  One less thing for new users to set up just helps the overall experience.

Maybe others have different opinions though? ping @kRITZCREEK @epost

kRITZCREEK commented:
I know that haskell keeps its flycheck support in a seperate package but I wouldn't mind putting it in here as long as there is an option to turn it off. If we could add that without making flycheck a dependency I would be even happier :)

epost commented:
Yeah, turning it off would be a good option to have.  I wouldn't mind the dependency either.

epost commented:
@bsermons, maybe it would be a good idea if you went ahead and stole some code from @bodil? We wouldn't have to tell her, necessarily.

bodil commented:
You're welcome to it!

bsermons commented:
Thanks! Updated the PR with @bodil's suggestion code.  Also, added you as an author if that is ok.

Not sure where the best place code to enable flycheck mode should go. Should I add a purescript-mode hook here or leave it up to the user to call `flycheck-mode`?

kRITZCREEK commented:
Code looks good to me, I'll give it a spin later today and make sure it also works with `0.9.1`.

kRITZCREEK commented:
I got this on the first load:

```emacs-lisp
File mode specification error: (error "Recursive load" "/home/creek/.emacs.d/elpa/psc-ide-20160602.1211/psc-ide.el" "/home/creek/.emacs.d/elpa/psc-ide-20160602.1211/psc-ide-flycheck.el" "/home/creek/.emacs.d/elpa/psc-ide-20160602.1211/psc-ide.el" "/home/creek/.emacs.d/elpa/psc-ide-20160602.1211/psc-ide-flycheck.el" "/home/creek/.emacs.d/elpa/psc-ide-20160602.1211/psc-ide.el" "/home/creek/.emacs.d/elpa/psc-ide-20160602.1211/psc-ide-flycheck.el" "/home/creek/.emacs.d/elpa/psc-ide-20160602.1211/psc-ide.el" "/home/creek/.emacs.d/elpa/psc-ide-20160602.1211/psc-ide-flycheck.el" "/home/creek/.emacs.d/elpa/psc-ide-20160602.1211/psc-ide.el")
```

bsermons commented:
Whoops, fixed now. Should have tested after closing/reopening emacs. :grin:

kRITZCREEK commented:
Okay now it started working. I just got this error though:

```
error in process sentinel: flycheck-same-files-p: Wrong type argument: stringp, nil
error in process sentinel: Wrong type argument: stringp, nil
```

bsermons commented:
Never seen that one.  Can you give more details on what you were doing or possibly steps to reproduce?

kRITZCREEK commented:
I was just playing around adding in some errors and removing them... I'll try something reproducable


Hmm seems like it stopped working after that again... Maybe that even was an error inside my async-send?

bsermons commented:
Ah yep, I can reproduce.  I'll have to investigate more tomorrow though.  Just disable and re-enable flycheck-mode fix it temporarily.

bsermons commented:
Ok, looked at it a bit more but not entirely sure what is going on still.  I am thinking it might be to do with  how spacemacs works when doing `:w` to save the file since it switches focus to the command bar at the bottom.  I have not reproduced it while saving the file with `C-x C-s` so that is what makes me think that.  Still needs some more investigating though.

I did have a problem where my error handling was and that was why it clobbered flycheck afterwards.  That is fixed with latest push so at least in the meantime the worst case is an error in the *Message* buffer for when it happens.

kRITZCREEK commented on psc-ide.el:
```diff
@@ -358,22 +358,25 @@ use when the search used was with `string-match'."
-                "start it using psc-ide-server-start."))))))
+  (let ((buffer (generate-new-buffer "*psc-ide-client*")))
+    (condition-case err
+        (let* (
```
Do we still need let* here?

bsermons replied:
Nah not really.  I need to fix that dangling open parenthesis too.

bsermons commented:
The problem was happening if you switched buffers before the send command was able to invoke the callback.  Should be fixed with the latest push since I now capture the current-buffer when the `psc-ide-send` is called use that buffer to evaluate the results in.

kRITZCREEK commented:
Sounds good :+1: I'll try it out later and merge!

bsermons commented:
Closing for https://github.com/epost/psc-ide-emacs/pull/48
